### PR TITLE
CommonGeometry_Engine-Issues423,856,860-CurveIntersections&IsSelfIntersecting

### LINKED ISSUE #423: Geometry_Engine: LineIntersection methods failing
### This issue has been archived 

Historic issue archive can be found [here](https://burohappold.sharepoint.com/sites/BHoM/01_Master%20File/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F01_Master%20File%2F999_BHoM_GitHub_History_Archive)

### LINKED ISSUE #856: Common_Engine: add IsSelfIntersecting method for IElement1D & IElement2D
Needed e.g. to resolve #723.

Depends on #724

### LINKED ISSUE #860: Geometry_Engine: Extend CurvePlanarIntersections to CurveIntersections

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #423 
Closes #856 
Closes #860 

### Test files

Available [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02_Current%2F12_Scripts%2F01_Test%20Scripts%2FBHoM_Engine%2FGeometry_Engine%2FCommonGeometry_Engine-Issues423%2C856%2C860-CurveIntersections%26IsSelfIntersecting) and LineIntersections.gh from [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02_Current%2F12_Scripts%2F01_Test%20Scripts%2FBHoM_Engine%2FGeometry_Engine%2FQuery) - in the latter a few cases fail, this is related to the tolerance (different approach in GH and BHoM).

### Changelog

- small change in logics of `LineIntersection(Line, Line)` and `LineIntersections(List<Line>)` methods:
    - the former now returns null when the lines are collinear (before one point was returned when the lines were overlapping - more confusing than helpful as usually there are 2 intersection points when the lines overlap)
    - the latter now returns up to 2 points for the overlapping lines - thanks to that o polyline consisting of overlapping lines is being considered self-intersecting
- `CurvePlanarIntersections` methods have been deprecated and extended to `CurveIntersections`
- `IsSelfIntersecting` methods added to curves, `IElement1D` and `IElement2D`

### Additional comments

@al-fisher @IsakNaslundBh please do not action before @LMarkowski reviews - I would go through the review process with him as an intro

### CODE REVIEW

File: Common_Engine/Query/IsSelfIntersecting.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2018, the respective contributors. All rights reserved.
```
Comment (IsakNaslundBh) [inline Common_Engine/Query/IsSelfIntersecting.cs]: Check the year here. Should be 2019 in new/modified files

File: Geometry_Engine/Query/CurvePlanarIntersections.cs [hunk 2]
```diff
@@ -33,149 +34,85 @@ public static partial class Query
...
-                iPts.Add(midPt - perp * shift);
-            }
-
-            return iPts;
+            return curve1.CurveIntersections(curve2, tolerance);
         }
 
         /***************************************************/
 
+        [DeprecatedAttribute("CurveIntersections")]
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Query/CurvePlanarIntersections.cs]: Please check the format of the constructor for the deprecated attribute here.

File: Common_Engine/Query/IsSelfIntersecting.cs (hunk 1 above)
Comment (pawelbaran) [reply Common_Engine/Query/IsSelfIntersecting.cs]: Fixed

File: Geometry_Engine/Query/CurvePlanarIntersections.cs (hunk 2 above)
Comment (pawelbaran) [reply Geometry_Engine/Query/CurvePlanarIntersections.cs]: Please have a look if you're happy with changes.

**Pull request merged**